File: python/완전탐색/피로도.py

""" Level 2. 피로도 """

def solution(k, dungeons):

    # 최소 필요 피로도 - 해당 던전을 탐험하기 위해 가지고 있어야 하는 최소한의 피로도
    # 소모 피로도 - 던전을 탐험한 후 소모되는 피로도
    
    global answer
    answer = 0

    length = len(dungeons)
    selected = [False] * length

    def move(total_hp, dcount):
        
        global answer
        answer = max(answer, dcount)

        for idx in range(length):
            if not selected[idx] and total_hp >= dungeons[idx][0]:
                selected[idx] = True
                move(total_hp - dungeons[idx][1], dcount+1)
                selected[idx] = False

    move(k, 0)

    return answer

print(solution(80, [[80,20],[50,40],[30,10]]))      # 3

File: Java/src/완전탐색/피로도.java
package 완전탐색;

public class 피로도 {

    static int answer;
    static int length;
    static boolean[] selected;
    
    public static void main(String[] args) {

        int k1 = 80;
        int[][] dungeons = {{80, 20}, {50, 40}, {30, 10}};
        System.out.println(solution(k1, dungeons));
    }   
    
    public static int solution(int k, int[][] dungeons) {
        
        answer = 0;
        length = dungeons.length;
        selected = new boolean[length];
        
        move(dungeons, k, 0);
        return answer;
    }
    
    public static void move(int[][] dungeons, int hp, int count) {
        
        answer = Math.max(count, answer);
    
        for (int i = 0; i < length; i++) {
            if (!selected[i] && dungeons[i][0] <= hp) {
                selected[i] = true;
                move(dungeons, hp-dungeons[i][1], count+1);
                selected[i] = false;
            }
        }
    }
}
